sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBubbleClockManager.scala

```diff
@@ -0,0 +1,205 @@
+
+  /**
+   * Periodical loop to check GPU utility and update clocks accordingly.
+   */
+  override def run(): Unit = {
+    while (true) {
+      val newVal = !GpuSemaphore.isDeviceBusy
+      if (isDeviceIdle != newVal) {
+        // Enter a critical section to update `clocks` and `isDeviceIdle` atomically
+        lock.lock()
+        val ts = System.nanoTime()
+        if (newVal) {
+          // GPU just exit a busy period, start all clocks
+          clocks.foreach { case (_, clk) => clk.tic(ts) }
+        } else {
+          // GPU just enter a busy period, stop all clocks
+          clocks.foreach { case (_, clk) => clk.toc(ts) }
+        }
+        isDeviceIdle = newVal
+        lock.unlock()
+      }
+      Thread.sleep(1L)
+    }
+  }
+
+  /**
+   * Register a new clock to the manager, a synchronized operation.
+   *
+   * @param clk the clock to be registered
+   * @return a unique id for the registered clock
+   */
+  def registerClock(clk: GpuBubbleWallClock): Long = {
+    lock.lock()
+    try {
+      // Sync the clock with current GPU status during initialization
+      clk.init(isDeviceIdle)
+      // Find a unique id for the new clock
+      var id = getNextId
+      while (clocks.contains(id)) {
+        id = getNextId
+      }
+      // Register the clock
+      clocks(id) = clk
+      id
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  /**
+   * Complete a clock and remove it from the manager, a synchronized operation.
+   *
+   * @param id the id of the clock to be completed
+   * @return the elapsed time in nanoseconds recorded by the clock
+   */
+  def completeClock(id: Long): Long = {
+    lock.lock()
+    try {
+      clocks.remove(id) match {
+        case None =>
+          throw new IllegalArgumentException(s"Invalid clock id: $id")
+        case Some(clk) =>
+          clk.finish()
+      }
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  // Lock to synchronize modifications to `clocks` and `isDeviceIdle`
+  private val lock = new ReentrantLock()
+
+  // Map to hold all registered clocks, only be accessed under lock
+  private val clocks = mutable.Map.empty[Long, GpuBubbleWallClock]
+
+  // Initialize the GPU status
+  private var isDeviceIdle: Boolean = !GpuSemaphore.isDeviceBusy
+
+  // This generator shall only be accessed under lock
+  private def getNextId: Long = {
+    val id = nextId
+    if (id == Long.MaxValue) {
+      nextId = 0L
+    } else {
+      nextId += 1
+    }
+    id
+  }
+
+  @volatile private var nextId: Long = 0L
+}
+
+/**
+ * A wall-clock timer that measures elapsed time during periods when the GPU is underloaded.
+ * The timer is designed to be managed by GpuBubbleClockManager.
+ */
+private class GpuBubbleWallClock {
+
+  def init(isDeviceIdle: Boolean): Unit = {
+    require(!initialized, "Timer was already initialized")
+    initialized = true
+    if (isDeviceIdle) {
+      startTime = System.nanoTime()
+    }
+  }
+
+  def tic(ts: Long): Unit = {
```

Comment:
If we go with `tic` and `toc`, then we need clear documentation on what each one is intended to do as it is not clear from the name.